src/picongpu/include/particles/creation/CreatorBase.hpp

```diff
@@ -0,0 +1,189 @@
+#include "fields/FieldE.hpp"
+
+#include "compileTime/conversion/TypeToPointerPair.hpp"
+#include "memory/boxes/DataBox.hpp"
+
+namespace picongpu
+{
+namespace particles
+{
+namespace creation
+{
+
+/** This functor is a base class for an actual particle creator.
+ *
+ * \tparam T_DestSpecies electron species to be created
+ */
+template<typename T_SourceSpecies, typename T_TargetSpecies>
+struct CreatorBase
+{
+
+    typedef T_SourceSpecies SourceSpecies;
+    typedef T_TargetSpecies TargetSpecies;
+
+    typedef typename SourceSpecies::FrameType FrameType;
+
+    /* specify field to particle interpolation scheme */
+    typedef typename PMacc::traits::Resolve<
+        typename GetFlagType<FrameType,interpolation<> >::type
+    >::type Field2ParticleInterpolation;
+
+    /* margins around the supercell for the interpolation of the field on the cells */
+    typedef typename GetMargin<Field2ParticleInterpolation>::LowerMargin LowerMargin;
+    typedef typename GetMargin<Field2ParticleInterpolation>::UpperMargin UpperMargin;
+
+    /* relevant area of a block */
+    typedef SuperCellDescription<
+        typename MappingDesc::SuperCellSize,
+        LowerMargin,
+        UpperMargin
+        > BlockArea;
+
+    //BlockArea BlockDescription;
+
+protected:
+
+    typedef MappingDesc::SuperCellSize TVec;
+
+    typedef FieldE::ValueType ValueType_E;
+    typedef FieldB::ValueType ValueType_B;
+    /* global memory EM-field device databoxes */
+    PMACC_ALIGN(eBox, FieldE::DataBoxType);
+    PMACC_ALIGN(bBox, FieldB::DataBoxType);
+    /* shared memory EM-field device databoxes */
+    PMACC_ALIGN(cachedE, DataBox<SharedBox<ValueType_E, typename BlockArea::FullSuperCellSize,1> >);
+    PMACC_ALIGN(cachedB, DataBox<SharedBox<ValueType_B, typename BlockArea::FullSuperCellSize,0> >);
+
+    /* host constructor initializing member : random number generator */
+    CreatorBase()
+    {
+        DataConnector &dc = Environment<>::get().DataConnector();
+        /* initialize pointers on host-side E-(B-)field databoxes */
+        FieldE* fieldE = &(dc.getData<FieldE > (FieldE::getName(), true));
+        FieldB* fieldB = &(dc.getData<FieldB > (FieldB::getName(), true));
+        /* initialize device-side E-(B-)field databoxes */
+        eBox = fieldE->getDeviceDataBox();
+        bBox = fieldB->getDeviceDataBox();
+    }
+
+    /** Initialization function on device
+     *
+     * \brief Cache EM-fields on device
+     *         and initialize possible prerequisites for ionization, like e.g. random number generator.
+     *
+     * This function will be called inline on the device which must happen BEFORE threads diverge
+     * during loop execution. The reason for this is the `__syncthreads()` call which is necessary after
+     * initializing the E-/B-field shared boxes in shared memory.
+     */
+    DINLINE void init(const DataSpace<simDim>& blockCell, const int& linearThreadIdx, const DataSpace<simDim>& localCellOffset)
```

Comment:
Please add a documentation for the params, localCellOffset can be anything and must described.